File: project/src/main/java/iteration2/project/CourseManager.java
package iteration2.project;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Random;

import org.json.simple.JSONArray;
import org.json.simple.JSONObject;

public class CourseManager {
	private static HashMap<String, ArrayList<Course>> curriculumCourses = new HashMap<String, ArrayList<Course>>();
	private static HashMap<String, ArrayList<Course>> electiveCourses = new HashMap<String, ArrayList<Course>>();
	private static ScheduleManager scheduleManager = new ScheduleManager(); // to schedule courses.
	Random rand = new Random();


	// create courses with schedule.
	public void createCourses() {
		// course attributes.
		String courseCode, courseName, prerequisiteCourse, courseType;
		int courseSemester, credit, akts, quota, practiceHour, theoricHour;
		Random rand = new Random(); // quota bilgisi eklenince silinecek.
		String[] courseCategory = { "semester1", "semester2", "semester3", "semester4", "semester5", "semester6",
				"semester7", "semester8", "FTE", "TE", "NTE" };

		/*** create courses and schedule. ***/

		// get semester courses.
		for (int i = 0; i < courseCategory.length; i++) {
			JSONArray courses = FileManager.readCourses(
					"C:\\Users\\murat\\eclipse-workspace\\project\\src\\main\\java\\iteration2\\iteration2\\project\\src\\main\\java\\jsonFiles\\courses.json",
					courseCategory[i]);
			// create and schedule course.
			// create
			for (Object c : courses) {
				courseCode = (String) ((JSONObject) c).get("Course Code");
				courseName = (String) ((JSONObject) c).get("Course Name");
				courseType = (String) ((JSONObject) c).get("Type");
				credit = (int) Float.parseFloat(((String) ((JSONObject) c).get("Credit")).replace(",", "."));
				akts = (int) Float.parseFloat(((String) ((JSONObject) c).get("ECTS")).replace(",", "."));
				quota = rand.nextInt(20) + 40;
				theoricHour = (int) Float.parseFloat(((String) ((JSONObject) c).get("T")).replace(",", "."));
				practiceHour = (int) Float.parseFloat(((String) ((JSONObject) c).get("U")).replace(",", "."));

				Course course;
				// action wrt. course type.
				switch (courseCategory[i]) {
				case "FTE":
					course = new FteCourse(i + 1, courseCode, courseName, credit, akts, quota, courseType);
					break;
				case "NTE":
					course = new NteCourse(i + 1, courseCode, courseName, credit, akts, quota, courseType);
					break;
				case "TE":
					course = new TeCourse(i + 1, courseCode, courseName, credit, akts, quota, courseType);
					break;

				default: // 'must' course.
					course = new Course(i + 1, courseCode, courseName, credit, akts, 1000, courseType);
					break;
				}

				// add prerequisite if exist.
				try {
					course.setPrerequisiteCourse((String) ((JSONObject) c).get("precondition"));
				} catch (Exception e) {
				}

				// schedule course.
				scheduleManager.scheduleCourse(course, practiceHour, theoricHour);

				// add course to hashMap.
				if (i < 8) { // curriculum courses
					if (curriculumCourses.get(courseCategory[i]) == null) {
						curriculumCourses.put(courseCategory[i], new ArrayList<Course>());
					}
					curriculumCourses.get(courseCategory[i]).add(course);
				} else { // elective courses.
					if (electiveCourses.get(courseCategory[i]) == null) {
						electiveCourses.put(courseCategory[i], new ArrayList<Course>());
					}
					electiveCourses.get(courseCategory[i]).add(course);
				}
			}
			scheduleManager.resetAvailableTimes();

		}

	}

	// get previous years' courses to create transcript
	public static ArrayList<Course> getPastSemesterCourses(int semester) {
		ArrayList<Course> result = new ArrayList<Course>();
		for (int i = 1; i < semester; i++) {
			ArrayList<Course> semesterCourses = new ArrayList<Course>(
					curriculumCourses.get("semester" + String.valueOf(i)));
			for (Course course : semesterCourses) {
				result.add(course);
			}
		}
		return result;
	}

	// get courses of requested semester for registration process.
	public static ArrayList<Course> getCurrentSemesterCourses(int semester) {
		return curriculumCourses.get("semester" + String.valueOf(semester));

	}

	// it prints courses with their days and hours.
	public void printSchedule() {
		for (int i = 1; i < 9; i++) {
			System.out.println("*************" + "SEMESTER " + i + "****************");
			for (Course course : curriculumCourses.get("semester" + i)) {
				int sectionNo = 0;
				System.out.println("********************");
				System.out.println("Course Name: " + course.getCourseName());
				System.out.println("Course Quota: " + course.getQuota());
				for (CourseSection section : course.getCourseSections()) {
					sectionNo++;
					System.out.println("Section " + sectionNo + ":" + section.getSchedule());

				}

			}
		}

	}

	
	public HashMap<String, ArrayList<Course>> getElectiveCourses(){
		return electiveCourses;
	}


	// select opened elective courses for this semester.
	public HashMap<String, ArrayList<Course>> getOpenedElectiveCourses(){
		HashMap<String, ArrayList<Course>> openedElectiveCourses = new HashMap<String, ArrayList<Course>>();
		int randIndex;
		for(int i=0; i<6; i++) {
			// TE
			randIndex = rand.nextInt(electiveCourses.get("TE").size());
			if (openedElectiveCourses.get("TE") == null) {
				openedElectiveCourses.put("TE", new ArrayList<Course>());
			}
			openedElectiveCourses.get("TE").add(electiveCourses.get("TE").get(randIndex));
			
			//FTE
			randIndex = rand.nextInt(electiveCourses.get("FTE").size());
			if (openedElectiveCourses.get("FTE") == null) {
				openedElectiveCourses.put("FTE", new ArrayList<Course>());
			}
			openedElectiveCourses.get("FTE").add(electiveCourses.get("FTE").get(randIndex));	
			
			//NTE
			randIndex = rand.nextInt(electiveCourses.get("NTE").size());
			if (openedElectiveCourses.get("NTE") == null) {
				openedElectiveCourses.put("NTE", new ArrayList<Course>());
			}
			openedElectiveCourses.get("NTE").add(electiveCourses.get("NTE").get(randIndex));	
			
		}
		
		return openedElectiveCourses;
	}

}

File: py_iteration/CourseManager.py
from Course import Course
from FileManager import FileManager
from FteCourse import FteCourse
from NteCourse import NteCourse
from ScheduleManager import ScheduleManager
import random

from TeCourse import TeCourse


class CourseManager:
    def __init__(self):
        self.curriculumCourses = {}  # key: semester, value: [Courses]
        self.electiveCourses = {}    # key: type, value:[Courses]
        self.scheduleManager = ScheduleManager()   # to create and schedule courses.

    def create_courses(self, course_categories):
        file_manager = FileManager.get_instance()
        for i in range(len(course_categories)):
            courses_json = file_manager.read_courses(course_categories[i])
            for obj in courses_json:
                quota_rand = random.randint(40, 70)
                # create course according to type.
                if course_categories[i] == 'FTE':
                    course = FteCourse(8, obj['Course Code'], obj['Course Name'], obj['Credit'], obj['ECTS'],
                                    quota_rand, obj['Type'])
                elif course_categories[i] == 'NTE':
                    course = NteCourse(8, obj['Course Code'], obj['Course Name'], obj['Credit'], obj['ECTS'],
                                       quota_rand, obj['Type'])
                elif course_categories[i] == 'TE':
                    course = TeCourse(8, obj['Course Code'], obj['Course Name'], obj['Credit'], obj['ECTS'],
                                      quota_rand, obj['Type'])
                else:  # must courses.
                    course = Course(i+1, obj['Course Code'], obj['Course Name'], obj['Credit'], obj['ECTS'],
                                    quota_rand, obj['Type'])

                # add prerequisite course if exists
                if 'precondition' in obj:
                    course.prerequisite_course = obj['precondition']

                # schedule the course
                theoretic_hour = int(obj['T'])
                practice_hour = int(obj['U'])
                self.scheduleManager.schedule_course(course, theoretic_hour, practice_hour)

                # add course to its course dict.
                if i < 8:  # must courses.
                    if course_categories[i] not in self.curriculumCourses:
                        self.curriculumCourses[course_categories[i]] = []
                    self.curriculumCourses[course_categories[i]].append(course)
                else:  # elective courses.
                    if course_categories[i] not in self.electiveCourses:
                        self.electiveCourses[course_categories[i]] = []
                    self.electiveCourses[course_categories[i]].append(course)

            # reset available times for next course category schedule
            self.scheduleManager.reset_available_times()

    # to create transcript BEFORE REGISTRATION.
    def get_past_semester_courses(self, semester):
        courses = []
        for i in range(1, semester):
            for course in self.curriculumCourses["semester" + str(i)]:
                courses.append(course)
        return courses

    def get_current_semester_courses(self, semester):
        return self.curriculumCourses[semester]

    # determine elective courses to open for registration.
    def get_five_elective_courses(self, elective_type):
        courses = []
        count = 0
        while count < 5:
            rand_index = random.randint(0, len(self.electiveCourses[elective_type]) - 1)
            if self.electiveCourses[elective_type][rand_index] not in courses:
                courses.append(self.electiveCourses[elective_type][rand_index])
                count += 1
        return courses

    def print_schedule(self):
        file = open('outputFiles/course_curriculum.txt', 'w')
        for semester in self.curriculumCourses:
            file.write("\n************{}************\n".format(semester))
            for course in self.curriculumCourses[semester]:
                file.write("Course Code: {}, Course Schedule: {}\n".format(course.course_code, course.course_sections[0].schedule))
        for typ in self.electiveCourses:
            file.write("\n************{}************\n".format(typ))
            for course in self.electiveCourses[typ]:
                file.write("Course Code: {}, Course Schedule: {}\n".format(course.course_code, course.course_sections[0].schedule))
        file.close()
